### PR TITLE
Perform division first for numerical stability

### PR DESCRIPTION
**Pull request by fredzzhang**
# Optimal Transport Layer

The current implementation of row/column normalisation is unstable when the mass of a given matrix is concentrated on a single row or column. Specifically, `r / alpha` and `c / beta` returns `inf` or `nan` for extremely small row/column sums. The corresponding row/column, however, has the same scale as the row/column sums. Thus, division should be executed first.

## Example
Data: [attn.pt.zip](https://github.com/anucvml/ddn/files/6860069/attn.pt.zip)
Code:
```python
>>> a = torch.load('attn.pt', map_location='cpu')
>>> # M is 1x12x12
>>> M = a[None, 4]
>>> H = W = 12
>>> B = 1
>>> c = r = 1.0 / H
>>> gamma = 1.0
>>> # Exponentiation
>>> P = torch.exp(-1.0 * gamma * (M - torch.amin(M, 2, keepdim=True)))
>>> # Row normalisation
>>> alpha = torch.sum(P, 2)
>>> P = (r / alpha).view(B, H, 1) * P
>>> # Column normalisation
>>> beta = torch.sum(P, 1)

>>> # Current implementation
>>> P_ = P * (c / beta).view(B, 1, W)
>>> print(torch.any(torch.isnan(P_)))
tensor(True)
>>> # Division first
>>> P__ = P / beta.view(B, 1, W) * c
>>> print(torch.any(torch.isnan(P__)))
tensor(False)
```

### CODE REVIEW

**Comment by sgould** [comment]
Check to make sure this works with non-default inputs for `r` and `c`, i.e., tensors of size (1,H) or (B,H) for `r` and (1,W) or (B,W) for `c`. Make sure it passes the tests at the end of the `optimal_transport.py` file.
